### PR TITLE
Federation E2E support on AWS via kops

### PR DESCRIPTION
**Pull request by colhom**
\cc @fejta @justinsb @kubernetes/sig-federation-pr-reviews

### CODE REVIEW

**Comment by colhom** [comment]
This bypasses the hacky `federation/cluster/*.sh` scripts and uses `kubefed` directly for provisioning the federation control plane on top of the kops clusters. 

\cc @madhusudancs

**Comment by madhusudancs** [comment]
I am a fan of replacing our shell scripts with code written in a real programming language that can be well tested. But to me the entire code in kubefed.go here feels like BASH written in Go, i.e. a Go wrapper around shell script. All it does is construct command line arguments and shell out. Is this necessary? It it worth it? `federation-up/down.sh` in their current form seem simpler and much easier to maintain which also essentially construct command line arguments and run the commands.

**Comment by colhom** [comment]
I don't see why we would build it like this:

 ` kubetest(go) --> federation_{up,down).sh (bash) --> kubefed (go)`. 

When we can have:

`kubetest (go) --> kubefed (go)`

Previously `kubetest` functionality was implemented in BASH, so I feel that there is already momentum within this context towards not using BASH wherever possible.

>  But to me the entire code in kubefed.go here feels like BASH written in Go

This is unavoidable, because `kubetest` and `kubefed` are compiled separately in the build process and hence `kubetest` will have to shell out to `kubefed` in some way.

In order to interact with the `federation_(up/down).sh` bash scripts, you would have to set a bunch of environment variables and shell out to the script, which is definitely messier than what we have here.

On a side note, do you happen to know what this bazel failure means?

```sh
************* Module scenarios.kubernetes_kops_aws
R: 66, 0: Too many statements (59/50) (too-many-statements)
Use --strategy=TestRunner=standalone to disable sandboxing for the failing actions.
```
@madhusudancs

**Comment by madhusudancs** [comment]
> When we can have:

> kubetest (go) --> kubefed (go)

But what is really happening here is: kubetest (go) --> kubetest(bash) --> kubefed (go)

> Previously kubetest functionality was implemented in BASH, so I feel that there is already momentum within this context towards not using BASH wherever possible.

This is true. But this uses BASH.

> This is unavoidable, because kubetest and kubefed are compiled separately in the build process and hence kubetest will have to shell out to kubefed in some way. 

Yeah. This is why it makes sense to make an exception and write that bridge in BASH instead of pretending that the bridge is in Go but in reality is Go-wrapped BASH.

> In order to interact with the federation_(up/down).sh bash scripts, you would have to set a bunch of environment variables and shell out to the script, which is definitely messier than what we have here.

Why? We have been doing this using env files that get sourced in the environment where the code runs. I agree setting individual variables is cumbersome, but not so much if you do wholesale. 

> ************* Module scenarios.kubernetes_kops_aws
> R: 66, 0: Too many statements (59/50) (too-many-statements)

This looks like a PyLint error that is saying your function has 59 statements whereas the limit is 50.

**Comment by colhom** [comment]
> Yeah. This is why it makes sense to make an exception and write that bridge in BASH instead of pretending that the bridge is in Go but in reality is Go-wrapped BASH.

There's no pretending here- a process spawning a sub-process and providing arguments does not in any way depend on BASH (or any shell interpreter). So there is really no BASH involved when a go binary exec's another go binary- you are just interacting with the operating system to make new sub-processes.

> Why? We have been doing this using env files that get sourced in the environment where the code runs. I agree setting individual variables is cumbersome, but not so much if you do wholesale.

It's detrimental for a variety of reasons.

*  I have literally seen our e2e grid held up multiple times by illegal BASH variable names or weird expansions that have an unescaped space or something like that.

*  There's no way to express interface arguments cleanly via structs or interfaces. Very often this leads to the case where you have to read and comprehend the entire BASH script before you know what env-var arguments it needs. This is definitely the case with our current scripts.

* Dealing with any data structure in BASH other than space-separated array is a real pain

* There's virtually no static validation you can do on a BASH script

All of these problems ( and more ) are trivially solved by pretty much any modern programming language. As this is a golang-shop, we should just use go for all control flow and be done with it.

**Comment by madhusudancs** [comment]
> There's no pretending here- a process spawning a sub-process and providing arguments does not in any way depend on BASH (or any shell interpreter). So there is really no BASH involved when a go binary exec's another go binary- you are just interacting with the operating system to make new sub-processes.

I did not mean this invokes the interpreter, I meant this was equivalent to doing that.

> I have literally seen our e2e grid held up multiple times by illegal BASH variable names or weird expansions that have an unescaped space or something like that.
>
> It's detrimental for a variety of reasons.
>
> *  I have literally seen our e2e grid held up multiple times by illegal BASH variable names or weird expansions that have an unescaped space or something like that.
> 
> *  There's no way to express interface arguments cleanly via structs or interfaces. Very often this leads to the case where you have to read and comprehend the entire BASH script before you know what env-var arguments it needs. This is definitely the case with our current scripts.
>
> * Dealing with any data structure in BASH other than space-separated array is a real pain
>
> * There's virtually no static validation you can do on a BASH script
> 
> All of these problems ( and more ) are trivially solved by pretty much any modern programming language. As this is a golang-shop, we should just use go for all control flow and be done with it.

All these are true and are solved if we are writing pure Go. But that's not what is happening here. We are still writing commands and command line arguments (language changed for clarity) which are still prone to some, if not all, of the problems but even more difficult to track because of Go-wrapping.

But anyway. I concede at this point and defer to @fejta

File: jenkins/e2e-image/kops-e2e-runner.sh [hunk 1]
```diff
@@ -52,15 +52,19 @@ if [[ "${KOPS_DEPLOY_LATEST_KUBE:-}" =~ ^[yY]$ ]]; then
...
 fi
 
-EXTERNAL_IP=$(curl -SsL -H 'Metadata-Flavor: Google' 'http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/access-configs/0/external-ip' || true)
-if [[ -z "${EXTERNAL_IP}" ]]; then
-  # Running outside GCE
-  echo
-  echo "WARNING: Getting external IP from instance metadata failed, assuming not running on GCE."
-  echo
-  EXTERNAL_IP=$(curl 'http://v4.ifconfig.co')
+if [[ "${E2E_OPT}" =~ "--kops-admin-access" ]];then
```
Comment (fejta) [inline jenkins/e2e-image/kops-e2e-runner.sh]: This is complicated logic. Please move it into `kubetest` and/or the scenario

File: kubetest/util.go [hunk 2]
```diff
@@ -168,6 +170,26 @@ func finishRunning(cmd *exec.Cmd) error {
...
+cmd: command to retry
+timeout: how long to retry for
+*/
+func retryFinishRunning(cmd *exec.Cmd, timeout time.Duration) error {
+	stepName := strings.Join(cmd.Args, " ")
+	var err error
+
+	for start := time.Now(); time.Since(start) < timeout; {
+		thisCmd := *cmd
+		err = finishRunning(&thisCmd)
```
Comment (fejta) [inline kubetest/util.go]: This needs to be aware of the `--timeout` flag

**Comment by fejta** [comment]
I don't have opinions about the goodness/badness of the `federation/cluster/*.sh` scripts. They lie beneath kubetest and so happy to defer to whoever is maintaining the federation stuff.

I do have opinions about all the junk above kubetest, and we should eliminate as much of this as possible. The goal is to eliminate the `e2e-runner.sh` scripts which first require us to make (and keep) these scripts trivial.

Note how the `scenarios/kubernetes_e2e.py --stage=foo` does not depend on injecting any environment variables, it just passes the args into the container: `runner_args.append('--stage=%s' % args.stage)` [link](https://github.com/kubernetes/test-infra/blob/38fbebb89ba54c4d1688518dc5a6d5e70f6a1e4d/scenarios/kubernetes_e2e.py#L307)

Can we do that here?

File: kubetest/util.go (hunk 2 above)
Comment (colhom) [inline kubetest/util.go]: The timeout is passed from the command line args [here](https://github.com/kubernetes/test-infra/pull/2576/files#diff-d1c748936b92e3673a2a848a2773a08eR161)

File: jenkins/e2e-image/kops-e2e-runner.sh (hunk 1 above)
Comment (colhom) [inline jenkins/e2e-image/kops-e2e-runner.sh]: This logic was already here- this PR simply adds a single-line conditional case to detect if the flag has already been specified and skip the external IP logic.

**Comment by colhom** [comment]
> I don't have opinions about the goodness/badness of the federation/cluster/*.sh scripts. They lie beneath kubetest and so happy to defer to whoever is maintaining the federation stuff.

I was the original author of the `federation/cluster` folder mess, and it was intended as a hacky intermediary solution when it was built. The goal (at the time) was to consolidate the logic spread across `e2e-runners`, `hack/e2e` and the `federation/cluster` folder into a single golang codebase. Now that we have `kubetest`, I don't think we should hesitate to consolidate both the `federation/cluster` logic and the `e2e-runner` logic into `kubetest`. 

> Note how the scenarios/kubernetes_e2e.py --stage=foo does not depend on injecting any environment variables, it just passes the args into the container: runner_args.append('--stage=%s' % args.stage) link

Agree that this would be an improvement, I just worry about this feature PR growing to refactor existing logic as well. I wonder if we can start putting together a list of things to change in the kops scenario/ e2e runner and have a follow PR? 

@fejta re: bazel failure. I assume it's telling me that the `main` function has > 50 statements? Should we really refactor the main function just so it has < 50 statements?

**Comment by colhom** [comment]
\cc @jdumars 

The `kubefed` API is available in `kubetest` via the `kubetest/kubefed.go` file added in this PR. If necessary, I can pull that file out into a discrete commit so you can cherry pick it easily. I think we should standardize on using `kubefed` in this way for now, as it's the most direct/least confusing way of provisioning federation control plane (thank you @madhusudancs  again for that!)

**Comment by jdumars** [comment]
@colhom thanks!

**Comment by colhom** [comment]
@k8s-bot pull-test-infra-bazel test this

**Comment by fejta** [comment]
/test all

Is this still alive? Please update and/or close

**Comment by colhom** [comment]
@fejta we decided not to go in this direction.

**Pull request closed without merging**